include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+                    physPg += miptailOffset;
+
+                    CPaddedCopyImageFilter::state_type copy;
+                    copy.outOffsetBaseLayer = (physPg).xyzz();/*physPg.z is layer*/ copy.outOffset.z = 0u;
+                    copy.inOffsetBaseLayer = core::vector2du32_SIMD(x,y)*m_pgSzxy;
+                    copy.extentLayerCount = core::vectorSIMDu32(m_pgSzxy, m_pgSzxy, 1u, 1u);
+                    copy.relativeOffset = {0u,0u,0u};
+                    if (x == w-1u)
+                        copy.extentLayerCount.x = std::max(extent.width>>(_subres.baseMipLevel+i),1u)-copy.inOffsetBaseLayer.x;
+                    if (y == h-1u)
+                        copy.extentLayerCount.y = std::max(extent.height>>(_subres.baseMipLevel+i),1u)-copy.inOffsetBaseLayer.y;
+                    memcpy(&copy.paddedExtent.width,(copy.extentLayerCount+core::vectorSIMDu32(2u*m_tilePadding)).pointer, 2u*sizeof(uint32_t));
+                    copy.paddedExtent.depth = 1u;
+                    if (w>1u)
+                        copy.extentLayerCount.x += m_tilePadding;
+                    if (x>0u && x<w-1u)
+                        copy.extentLayerCount.x += m_tilePadding;
+                    if (h>1u)
+                        copy.extentLayerCount.y += m_tilePadding;
+                    if (y>0u && y<h-1u)
+                        copy.extentLayerCount.y += m_tilePadding;
+                    if (x == 0u)
+                        copy.relativeOffset.x = m_tilePadding;
+                    else
+                        copy.inOffsetBaseLayer.x -= m_tilePadding;
+                    if (y == 0u)
+                        copy.relativeOffset.y = m_tilePadding;
+                    else
+                        copy.inOffsetBaseLayer.y -= m_tilePadding;
+                    copy.inMipLevel = _subres.baseMipLevel + i;
+                    copy.outMipLevel = 0u;
+                    copy.inImage = _img;
+                    copy.outImage = storage->image.get();
+                    copy.axisWraps[0] = _wrapu;
+                    copy.axisWraps[1] = _wrapv;
+                    copy.axisWraps[2] = ISampler::ETC_CLAMP_TO_EDGE;
+                    copy.borderColor = _borderColor;
+                    if (!CPaddedCopyImageFilter::execute(&copy))
+                        _IRR_DEBUG_BREAK_IF(true);
+                }
+        }
+
+        return offsetToTextureData(pgtOffset, _img, _wrapu, _wrapv);
+    }
+
+    bool free(const STextureData& _addr, const IImage* _img, const IImage::SSubresourceRange& _subres) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return false;
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+        }
```

Comment:
this is gonna be common to both IGPU and ICPU VT, the types will just be different, factor it out